python/lsst/pipe/base/connections.py

```diff
@@ -607,3 +629,111 @@ def iterConnections(connections: PipelineTaskConnections,
+                tuple(self.outputs.get(dataset_type_name, ()))
+            )
+        # Actually call adjustQuantum.
+        adjusted_inputs_by_connection, adjusted_outputs_by_connection = connections.adjustQuantum(
+            inputs_by_connection,
+            outputs_by_connection,
+            label,
+            data_id,
+        )
+        # Translate adjustments to DatasetType-keyed, Quantum-oriented form,
+        # installing new mappings in self if necessary.
+        adjusted_inputs: typing.Optional[NamedKeyDict[DatasetType, typing.List[DatasetRef]]] = None
+        adjusted_outputs: typing.Optional[NamedKeyDict[DatasetType, typing.List[DatasetRef]]] = None
+        for name, (connection, updated_refs) in adjusted_inputs_by_connection.items():
+            dataset_type_name = connection.name
+            if not set(updated_refs).issubset(self.inputs[dataset_type_name]):
+                raise RuntimeError(
+                    f"adjustQuantum implementation for task with label {label} returned {name} "
+                    f"({dataset_type_name}) input datasets that are not a subset of those "
+                    f"it was given for data ID {data_id}."
+                )
+            if adjusted_inputs is None:
+                adjusted_inputs = NamedKeyDict(self.inputs)
+            adjusted_inputs[dataset_type_name] = list(updated_refs)
+        for name, (connection, updated_refs) in adjusted_outputs_by_connection.items():
+            if not set(updated_refs).issubset(self.outputs[dataset_type_name]):
+                raise RuntimeError(
+                    f"adjustQuantum implementation for task with label {label} returned {name} "
+                    f"({dataset_type_name}) output datasets that are not a subset of those "
+                    f"it was given for data ID {data_id}."
+                )
+            if adjusted_outputs is None:
+                adjusted_outputs = NamedKeyDict(self.outputs)
+            adjusted_outputs[dataset_type_name] = list(updated_refs)
+        if adjusted_inputs is not None:
+            self.inputs = adjusted_inputs.freeze()
+            self.inputs_adjusted = True
+        else:
+            self.inputs_adjusted = False
+        if adjusted_outputs is not None:
+            self.outputs = adjusted_outputs.freeze()
+            self.outputs_adjusted = True
+        else:
+            self.outputs_adjusted = False
+
+    inputs: NamedKeyMapping[DatasetType, typing.List[DatasetRef]]
+    """Mapping of regular input and prerequisite input datasets, grouped by
+    `DatasetType`.
+    """
+
+    outputs: NamedKeyMapping[DatasetType, typing.List[DatasetRef]]
+    """Mapping of output datasets, grouped by `DatasetType`.
+    """
+
+    inputs_adjusted: bool = False
+    """Whether any inputs were removed in the last call to `adjust_in_place`.
+    """
+
+    outputs_adjusted: bool = False
+    """Whether any outputs were removed in the last call to `adjust_in_place`.
+    """
```

Comment:
🤷.  I am always uncertain how to order Python attributes in classes, so if there is any kind of standard, please point me at it.

Reply:
I think pep 557 where they are introduced is as authoritative as a source as any, since this was the proposal people accepted, and in all the examples they are listed first.